src/components/cylc/tree/TreeItem.vue

```diff
@@ -382,26 +382,20 @@ export default {
       traverseChildren(this.node)
       return tasks
     },
-    // this has be used in conjunction with branchingLineage since it will always return the first child otherwise
-    lastDescendent () {
+    /** Last descendant node that is the only child of its parent. */
+    lastSingleDescendant () {
```

Comment:
Rather than having this as a computed property of the TreeItem and passing that property into the template, could we have it as a method of the GScan component and call it on the (reactive) node?

That would keep the GScan logic in the GScan component, avoiding the overheads on the tree view.

Reply:
Not completely straightforward as other computed properties depend on this one. And I think the caching of computed properties is desirable. So the only real solution I can think of is to split this component in two - a GScan tree item and tree view tree item (#1417)

Reply:
If one computed property depends on another then they can be combined into a single function?

Isn't this:

```js
myFunction(reactiveProperty)  // gets recomputed when reactiveProperty changes
```

Essentially equivalent to this:

```js
myDerivedProperty // gets recomputed when reactiveProperty changes
// ...
computed: { myDerivedProperty () { return myFunction(reactiveProperty) } }
```

There are a couple of other solutions too:

1. Solve this at the node level (keeps the logic out of the tree component).
2. Use a mixin (avoids splitting the tree component).

Reply:
Computed properties only get re-evaluated when their reactive dependencies change, whereas methods get re-evaluated on every re-render, see https://vuejs.org/guide/essentials/computed.html#computed-caching-vs-methods

Looking at `TreeItem.vue` there's actually very little in common between the tree view related stuff and the GScan related stuff, hence why I think splitting is the best option. But we can certainly use a mixin (or composable) for the common stuff.

Reply:
Good to know on the reactivity front.

IMO, there should be a simple abstract `Tree` component which just does tree thing, like indentation, recursion, expand-collapse, etc. GScan and the Tree view should both use this Tree component. We should be able to swap for another off-the-shelf tree component if it is desirable to do so e.g. [the Element virtualised Tree](https://element-plus.org/en-US/component/tree-v2.html). Unfortunately the logic has leaked and the Tree component is far from abstract.

Note we can solve this at the node level as suggested above, the GScan component is already defining its own node tree. We should be able to change the data going into the tree to reflect the state-totals trickling stuff.

Reply:
(Copied some of this to #1417, suggest we continue discussion there)

Reply:
I'm a bit concerned about the possible performance impact on the tree view, especially given that the current performance is insufficient for some examples.